controllers/clustertopology_controller_compute_test.go

```diff
@@ -250,7 +252,281 @@ func TestGetClass(t *testing.T) {
+	machineDeploymentNoInfrastructure := newFakeMachineDeployment(metav1.NamespaceDefault, "no-infra").withLabels(labelsInClass).withBootstrapRef(objToRef(machineDeploymentBootstrap)).Obj()
+	machineDeploymentOutsideCluster := newFakeMachineDeployment(metav1.NamespaceDefault, "wrong-cluster-label").withLabels(labelsNotInClass).withBootstrapRef(objToRef(machineDeploymentBootstrap)).withInfraTemplateRef(objToRef(machineDeploymentInfrastructure)).Obj()
+	machineDeploymentUnmanaged := newFakeMachineDeployment(metav1.NamespaceDefault, "no-managed-label").withLabels(labelsUnmanaged).withBootstrapRef(objToRef(machineDeploymentBootstrap)).withInfraTemplateRef(objToRef(machineDeploymentInfrastructure)).Obj()
+
+	emptyMachineDeployments := make([]machineDeploymentTopologyState, 0)
+	tests := []struct {
+		name    string
+		cluster *clusterv1.Cluster
+		class   clusterv1.ClusterClass
+		objects []client.Object
+		want    *clusterTopologyState
+		wantErr bool
+	}{
+		{
+			name:    "Cluster exists with no references",
+			cluster: newFakeCluster(metav1.NamespaceDefault, "cluster1").Obj(),
+			// Expecting valid return with no control plane or infrastructure state defined and empty machine deployment state list
+			want: &clusterTopologyState{
+				cluster:               newFakeCluster(metav1.NamespaceDefault, "cluster1").Obj(),
+				controlPlane:          controlPlaneTopologyState{},
+				infrastructureCluster: nil,
+				machineDeployments:    emptyMachineDeployments,
+			},
+		},
+		{
+			name:    "Cluster with non existent infrastructure reference only",
+			cluster: newFakeCluster(metav1.NamespaceDefault, "cluster1").withInfraRef(objToRef(nonExistentInfraCluster)).Obj(),
+			objects: []client.Object{
+				infraCluster,
+			},
+			wantErr: true, // this test fails as partial reconcile is undefined.
+		},
+		{
+			name:    "Cluster infrastructure reference only",
+			cluster: newFakeCluster(metav1.NamespaceDefault, "cluster1").withInfraRef(objToRef(infraCluster)).Obj(),
+			objects: []client.Object{
+				infraCluster,
+			},
+			// Expecting valid return with no control plane or machine deployment state defined but with a valid infrastructure state
+			want: &clusterTopologyState{
+				cluster:               newFakeCluster(metav1.NamespaceDefault, "cluster1").withInfraRef(objToRef(infraCluster)).Obj(),
+				controlPlane:          controlPlaneTopologyState{},
+				infrastructureCluster: infraCluster,
+				machineDeployments:    emptyMachineDeployments,
+			},
+		},
+		{
+			name:    "Cluster with infrastructure reference and control plane reference, no control plane infrastructure and a clusterClass with no infra requirement",
+			cluster: newFakeCluster(metav1.NamespaceDefault, "cluster1").withControlPlaneRef(objToRef(controlPlane)).withInfraRef(objToRef(infraCluster)).Obj(),
```

Comment:
I would really prefer if those builders would be written multi-line. At least for me it's harder to read with those long lines. But maybe that's just me.

Reply:
Same applies (at least for me) to long inline struct definitions instead of multi-line (e.g. l.407)

Reply:
I think I'll leave this for now - builders are currently being reworked in #5055